crates/transpiler/src/transpile_layout.rs

```diff
@@ -0,0 +1,493 @@
+        } else {
+            (0..self.output_qubit_count)
+                .map(|x| {
+                    self.initial_layout
+                        .virtual_to_physical(VirtualQubit::new(x))
+                })
+                .collect()
+        }
+    }
+
+    /// Generate a routing permutation as an array of PhysicalQubit indices.
+    pub fn routing_permutation(&self) -> Vec<PhysicalQubit> {
+        match self.routing_permutation {
+            Some(ref perm) => perm.clone(),
+            None => (0..self.output_qubit_count)
+                .map(PhysicalQubit::new)
+                .collect(),
+        }
+    }
+
+    /// Generate the final layout as an array of PhysicalQubits
+    ///
+    /// This method will generate an array of final positions for each qubit in the input circuit.
+    /// For example, if you had an input circuit like:
+    ///
+    /// ```python
+    /// qc = QuantumCircuit(3)
+    /// qc.h(0)
+    /// qc.cx(0, 1)
+    /// qc.cx(0, 2)
+    /// ```
+    ///
+    /// and then the output from the transpiler was:
+    ///
+    /// ```python
+    /// tqc = QuantumCircuit(3)
+    /// tqc.h(2)
+    /// tqc.cx(2, 1)
+    /// tqc.swap(0, 1)
+    /// tqc.cx(2, 1)
+    /// ```
+    ///
+    /// then the `final_layout` method returns:
+    ///
+    /// ```python
+    /// [2, 0, 1]
+    /// ```
+    ///
+    /// This can be seen as follows. Qubit 0 in the original circuit is mapped to qubit 2
+    /// in the output circuit during the layout stage, which is mapped to qubit 2 during the
+    /// routing stage. Qubit 1 in the original circuit is mapped to qubit 1 in the output
+    /// circuit during the layout stage, which is mapped to qubit 0 during the routing
+    /// stage. Qubit 2 in the original circuit is mapped to qubit 0 in the output circuit
+    /// during the layout stage, which is mapped to qubit 1 during the routing stage.
+    /// The output list length will be as wide as the input circuit's number of qubits,
+    /// as the output list from this method is for tracking the permutation of qubits in the
+    /// original circuit caused by the transpiler.
+    ///
+    /// # Args
+    ///     `filter_ancillas` - If set to `true` any ancilla qubits added to
+    ///     the circuit by the transpiler will not be included in the output
+    ///     array.
+    pub fn final_layout(&self, filter_ancillas: bool) -> Vec<PhysicalQubit> {
+        let qubit_range = if filter_ancillas {
+            0..self.input_qubit_count
+        } else {
+            0..self.output_qubit_count
+        };
+        qubit_range
+            .map(|idx| {
+                let mut qubit_idx = self
+                    .initial_layout
+                    .virtual_to_physical(VirtualQubit::new(idx));
+                if let Some(ref routing_permutation) = self.routing_permutation {
+                    qubit_idx = routing_permutation[qubit_idx.index()]
+                }
+                qubit_idx
+            })
+            .collect()
+    }
```

Comment:
Please let's not make the same mistake as the Python struct, and instead return a clean `NLayout` that corresponds to the Python struct's `TranspileLayout.final_virtual_layout` / `TranspileLayout.final_index_layout` with the `VirtualQubit`s being the incoming qubits.  I think that's the same as the vector that comes out of this, but it's not possible to misinterpret because of the typing.

Reply:
I made 2 methods here one for an NLayout and the other for the vec return in: https://github.com/Qiskit/qiskit/pull/14778/commits/464dcca24f6cc8967145e5c31865dc1d9c384cd3